### ISSUE TITLE
"watch/builds/<build_id>/" returns states for all builds on the namespace

### ISSUE THREAD

**Issue by TomasTomecek** [openshift/origin#1634]
My build has finished and master is producing inconsistent results:

```
GET /osapi/v1beta1/watch/builds/rsyslog-docker-20150408-103244/?namespace=default HTTP/1.1
{u'object': {u'status': u'Failed', u'kind': u'Build', u'parameters': ..., u'type': u'MODIFIED'}
```

and

```
GET /osapi/v1beta1/builds/rsyslog-docker-20150408-103244/?namespace=default HTTP/1.1
{u'status': u'Complete', u'kind': u'Build', u'parameters': ...,
```

`watch` is of course incorrect since the container exited with 0:
```
a071d8c41bfe        buildroot:latest              "dock --verbose insi   4 minutes ago       Exited (0) 2 minutes ago                       k8s_custom-build.ec860de7_build-rsyslog-docker-20150408-103244.default.api_d9d1bc2d-ddc9-11e4-a464-6cae8b5467ca_ab986f0d
```

My workflow:
 1. create build: `POST /builds/`
 2. during build, get build json `GET /builds/<build_id>`, add metadata to labels and store the build json: `PUT /builds/<build_id>`
 3. watch the build on client (`GET /watch/builds/<build_id>`) and once it's done, return (client reports that build failed due to watch, while `GET /builds/<build_id>` returns that it completed well)

**Comment by csrwng** [comment]
@TomasTomecek what you are seeing is the state of a different build. Currently the watch will return the state of all builds on that namespace and not just the build you're requesting. This will be resolved when we switch the builds storage to use the upstream etcd generic controller. One workaround you can do for now is look at the object you're getting with the watch and check that the name equals the one that you're querying.

**Comment by TomasTomecek** [comment]
Wow, totally missed that. Thanks Cesar!

**Comment by TomasTomecek** [comment]
When do you plan to change the controller?

**Comment by TomasTomecek** [comment]
@csrwng ^

**Comment by csrwng** [comment]
@TomasTomecek I am working on making that change this week.

**Comment by liggitt** [comment]
@csrwng did the build registry get converted yet?

**Comment by csrwng** [comment]
I did not, because I found out that wasn't the source of the issue and other things took precedence. We need to have label conversion functions in the build API code.

**Comment by csrwng** [comment]
@TomasTomecek you can filter a build watch by adding a fields parameter that includes a name condition. Example:
```
https://localhost:8443/osapi/v1beta1/builds?namespace=myproject&watch=true&fields=name%3Dmybuild
```
The fields parameter must include a URL-encoded field filter condition. In the above case, the condition is name=mybuild

**Comment by TomasTomecek** [comment]
Is this a new api call? If so, since what version it is present?

**Comment by csrwng** [comment]
@TomasTomecek, you can still use the same watch call you've been using... I used the newer form above which uses a watch parameter on the GET of a collection, but it's not necessary for the filter to work. This one works just as well:
```
https://localhost:8443/osapi/v1beta1/watch/builds?namespace=myproject&fields=name%3Dmybuild
```

**Comment by TomasTomecek** [comment]
@csrwng Okay, thanks Cesar!

**Comment by mmilata** [comment]
Confirming fixed in 0.5.2.

**Comment by TomasTomecek** [comment]
Also verified; thanks Cesar!
